UofT-Frosh-Orientation/orientation-website
# 612 registration formik

## **Contribution:**

Refactoring registration form
- made changes to checkbox, radio and text input so they use the proper "tag form" (i.e., onChange)
  - no more warnings (yay!) 
- wrapped all components within form -- making them tabbable (improving accessibility)

Made changes to the button component
- also allowing tabbing and focus (improving accessibility) 

Made changes to dropdown component
- added keyboard focus and `onKeyDown` events (so you can tab through the options and press enter to select when focused)

Made changes to button component
- now keyboard focusable (tabs and enter)

Fixed some other small bugs with registration disable fields
- found a small bug that will require a lot of refactoring (not that important though?)
  - if you click Yes commuting, and then Subway (for example), the text box is enabled, but later if you click No not commuting, and click back to Yes, the text box is not enabled
  - this is because we only disable onChange, and since the dropdown with Subway isn't "changing" the textbox is not enabled

### **Issue:**

Closes:
- #612 
- #333 (hopefully fixed? added some local storage stuff)
- #330 (fixed some disabling field logic)
- #619

### **Accomplishments:**

- technically you can fill out the entire form without clicking anything (radio buttons are HTML tags and use arrow keys to change, checkboxes use enter to select)

### **Visuals:** 

No visuals changed

## **Details:**

### **Expected Behavior:**

- behaves exactly the same in terms of form submission (all fields in the frosh user are still there)
- clicking Tab will allow you to move through all the form fields
- clicking enter on the last tab for "Pay Now" will take you to the payment page (clicking also works as well)
- button is focusable (if you go to inspect tab you will be able to see keyboard-focusable with a green checkmark)
- bug in #330 is fixed

### **Testing Steps:**

Steps required to get expected behavior.

1. create a new account, and fill in all the fields
2. click the button to head over to the registration page and try to fill in all the fields just by using the keyboard
  - since some of the components are HTML tags which have their own functionality (some stuff might be confusing but try it out)
3. test out the [following](https://github.com/UofT-Frosh-Orientation/orientation-website/issues/330) on the Extra Events page
4. submit registration form on last page, and check db to make sure user was properly created

### **Complications:**

- `:focus` is the style when you're focused on that element -- it has inherent style so set `outline: none` in css to customize
- warning where it says something about not unique key -- try using `React.Fragment` tag and put the key into that
- more complications encountered, and solved with links in resources

### **Resources:**

Any additional resources used outside this in issue.

- `<React.Fragment>`: Using <React.Fragment>, you can create a container that doesn't render any additional elements when rendered. It acts as an invisible wrapper and doesn't create any extra nodes in the DOM hierarchy. `key` prop can be applied to the `<React.Fragment>` element surrounding the label and input elements. This ensures that each fragment has a unique key within the loop.
- [Radiobuttons](https://stackoverflow.com/questions/68419302/radio-button-in-react-js-using-map): you need to give each input a unique name attribute, as that's what groups them together. make sure the name you give doesn't repeat, otherwise, you'll have multiple radio groups controlling the same value.
- Making components [keyboard-focusable](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Keyboard)
- using [onKeyDown](https://www.tutorialspoint.com/detecting-arrow-key-presses-in-javascript)
